android/modules/network/src/java/ti/modules/titanium/network/socket/TCPProxy.java

```diff
@@ -220,22 +242,74 @@ public void run()
 			Object timeoutProperty = getProperty("timeout");
 
 			try {
-				if (timeoutProperty != null) {
-					int timeout = TiConvert.toInt(timeoutProperty, 0);
-
+				int port = TiConvert.toInt(getProperty("port"));
+				boolean useTls = TiConvert.toBoolean(getProperty("useTls"), false);
+
+				if (useTls) {
+					TiSocketFactory sslSocketFactory = new TiSocketFactory(null, null, NetworkModule.TLS_DEFAULT);
+					SSLSocket sslSocket = (SSLSocket) sslSocketFactory.createSocket();
+					sslSocket.setUseClientMode(true);
+					if (Build.VERSION.SDK_INT >= 24) {
+						SSLParameters sslParameters = new SSLParameters();
+						List sniHostNames = new ArrayList(1);
+						sniHostNames.add(new SNIHostName(host));
+						sslParameters.setServerNames(sniHostNames);
+						sslSocket.setSSLParameters(sslParameters);
+					}
+					clientSocket = sslSocket;
+				} else {
 					clientSocket = new Socket();
-					clientSocket.connect(new InetSocketAddress(host, TiConvert.toInt(getProperty("port"))), timeout);
+				}
 
+				InetSocketAddress endpoint = new InetSocketAddress(host, port);
+				if (timeoutProperty != null) {
+					int timeout = TiConvert.toInt(timeoutProperty, 0);
+					clientSocket.connect(endpoint, timeout);
 				} else {
-					clientSocket = new Socket(host, TiConvert.toInt(getProperty("port")));
+					clientSocket.connect(endpoint);
 				}
-				updateState(SocketModule.CONNECTED, "connected", buildConnectedCallbackArgs());
 
+				if (useTls) {
+					SSLSocket sslSocket = (SSLSocket) clientSocket;
+					SSLSession sslSession = sslSocket.getSession();
+					KrollFunction checkServerIdentity = (KrollFunction) getProperty("checkServerIdentity");
+					if (checkServerIdentity != null) {
+						Log.d(TAG, "checkServerIdentity");
+						X500Principal peerPrincipal = (X500Principal)sslSession.getPeerPrincipal();
+						String canonicalName = peerPrincipal.getName(X500Principal.CANONICAL);
+						int separatorIndex = canonicalName.indexOf(",");
+						String commonName = canonicalName.substring(canonicalName.indexOf("cn=") + 3, separatorIndex != -1 ? separatorIndex : canonicalName.length());
+						HashMap<String, Object> certObject = new HashMap();
+						HashMap<String, String> subjectMap = new HashMap();
+						subjectMap.put("CN", commonName);
+						certObject.put("subject", subjectMap);
+						Object result = checkServerIdentity.call(getKrollObject(), new Object[] { host, certObject });
```

Comment:
@jquick-axway There is some strange behavior going on here. If this callback returns an `Error` object, i get an empty `HashMap` as the result. Not sure if this is related due to this running on a separate thread. Only happens on Android, iOS works fine.

```js
checkServerIdentity(hostname, cert) {
  return new Error('failed'); // will result in an empty HashMap
  // return {message: 'failed'} // will result in a properly populated HashMap
}
```

Reply:
I think I know what's wrong. In V8, the JavaScript `Error` object's properties are not enumerable. I had to modify the `Error` prototype below so that `JSON.stringify()` could retrieve the non-enumerable properties.
https://github.com/appcelerator/titanium_mobile/blob/master/common/Resources/ti.internal/extensions/js/Error.js

Reply:
Ah yeah that explains it. Thanks for all the details! I wanted to have the API close to what Node.js does hence using the error as return value. I'll try a few other variations and update the PR once i settled for one.